allennlp/tests/training/gan_callback_trainer_test.py

```diff
@@ -0,0 +1,439 @@
+            fake_data = trainer.model.generator(array)
+            output = trainer.model.discriminator(fake_data["output"], torch.zeros(1))
+            self.loss = output["loss"]
+            self.discriminator_fake_loss += self.loss.sum().item()
+        elif stage == "generator":
+            generated = trainer.model.generator(array, trainer.model.discriminator)
+            fake_data = generated["output"]
+            self.loss = generated["loss"]
+            self.generator_loss += self.loss.sum().item()
+
+            self.fake_mean += fake_data.mean()
+            self.fake_stdev += fake_data.std()
+            self.count += 1
+
+    @handle_event(Events.BACKWARD)
+    def backpropagate_errors(self, trainer):
+        self.loss.backward()
+        trainer.train_loss += self.loss.item()
+
+    @handle_event(Events.AFTER_BACKWARD)
+    def optimize(self, trainer):
+        # pylint: disable=no-self-use
+        trainer.optimizer.step()
+
+    @handle_event(Events.BATCH_END)
+    def compute_metrics(self, trainer):
+        trainer.train_metrics = {
+            "dfl": self.discriminator_fake_loss,
+            "drl": self.discriminator_real_loss,
+            "gl": self.discriminator_real_loss,
+            "mean": self.fake_mean / max(self.count, 1),
+            "stdev": self.fake_stdev / max(self.count, 1)
+        }
+
+# This is bad, I have to monkeypatch Optimizer.from_params
+_original_optimizer_from_params = Optimizer.from_params  # pylint: disable=invalid-name
+
+def _optimizer_from_params(model_parameters: List, params: Params) -> Optimizer:
+    typ3 = params.get("type")
+    if typ3 == "gan":
+        generator_parameters = [[n, p] for n, p in model_parameters if n.startswith("generator.")]
+        discriminator_parameters = [[n, p] for n, p in model_parameters if n.startswith("discriminator.")]
+
+        return GanOptimizer(
+                generator_optimizer=_original_optimizer_from_params(generator_parameters,
+                                                                    params.pop("generator_optimizer")),
+                discriminator_optimizer=_original_optimizer_from_params(discriminator_parameters,
+                                                                        params.pop("discriminator_optimizer"))
+        )
+    else:
+        return _original_optimizer_from_params(model_parameters, params)
+
+# This is also bad, I have to provide a do-nothing dataset reader.
+@DatasetReader.register("nil")
+class NilDatasetReader(DatasetReader):
+    # pylint: disable=abstract-method
+    def _read(self, file_path):
+        return [Instance({})]
+
+
+def config(sample_size: int = 500,
+           batches_per_epoch: int = 40,
+           num_epochs: int = 50,
+           learning_rate: float = 0.05) -> Params:
+    return Params({
+        # These are unnecessary but the TrainerPieces.from_params expects them
+        "iterator": {"type": "basic"},
+        "dataset_reader": {"type": "nil"},
```

Comment:
Could you get around this by defining a `GanTrainer(CallbackTrainer)` that basically just defines a new constructor and calls `super().__init__`?  No other overrides necessary?  That way you could probably also have the constructor take two optimizers and two models, and not need the other hacks that you have.  Except the `**extras` stuff would depend on having a single named `model` argument that gets passed to each optimizer, so you would still need a custom `from_params`...  But that would at least let you isolate the problem to `GanTrainer.from_params`, without the need for a kludgey model and optimizer grouping.

The issue is that the types get a little off in the callback.  But it looks like they already are, because you don't have type annotations on `trainer` in your `TrainGan` callback.  This way, you could annotate it as `GanTrainer`, which is a subclass of `CallbackTrainer`, and things should be fine (except for the `handle_event` decorator, but I don't think mypy is sophisticated enough to check that anyway).